libraries/chain/include/scorum/chain/services/budgets.hpp

```diff
@@ -0,0 +1,135 @@
+#pragma once
+
+#include <scorum/chain/services/service_base.hpp>
+#include <scorum/chain/schema/budget_objects.hpp>
+
+#include <boost/lambda/lambda.hpp>
+#include <boost/multi_index/detail/unbounded.hpp>
+
+namespace scorum {
+namespace chain {
+
+struct fund_budget_service_i : public base_service_i<fund_budget_object>
+{
+};
+
+using dbs_fund_budget = dbs_service_base<fund_budget_service_i>;
+
+template <class ObjectType> struct owned_base_budget_service_i : public base_service_i<ObjectType>
+{
+    using budget_cref_type = typename base_service_i<ObjectType>::object_cref_type;
+
+    using budgets_type = std::vector<budget_cref_type>;
+
+    virtual const ObjectType& get(const typename ObjectType::id_type&) const = 0;
+    virtual budgets_type get_budgets() const = 0;
+    virtual budgets_type get_budgets_by_start_time(const fc::time_point_sec& until) const = 0;
+    virtual std::set<std::string> lookup_budget_owners(const std::string& lower_bound_owner_name,
+                                                       uint32_t limit) const = 0;
+    virtual budgets_type get_budgets(const account_name_type& owner) const = 0;
+};
+
+template <class InterfaceType, class ObjectType> class dbs_owned_base_budget : public dbs_service_base<InterfaceType>
+{
+    friend class dbservice_dbs_factory;
+
+protected:
+    explicit dbs_owned_base_budget(database& db)
+        : dbs_service_base<InterfaceType>(db)
+    {
+    }
+
+public:
+    const ObjectType& get(const typename ObjectType::id_type& id) const override
+    {
+        try
+        {
+            return this->get_by(id);
+        }
+        FC_CAPTURE_AND_RETHROW((id))
+    }
+
+    typename InterfaceType::budgets_type get_budgets() const override
```

Comment:
Define this `typename InterfaceType::budgets_type` in the scope of this class like `using budgets_type = typename InterfaceType::budgets_type` and use it instead